custom_components/solcast_solar/solcastapi.py

```diff
@@ -1915,69 +1990,107 @@ async def __http_data_call(self, site=None, api_key=None, do_past=False, force=F
-                            "pv_estimate10": x["pv_estimate10"],
-                            "pv_estimate90": x["pv_estimate90"],
+                            "pv_estimate": forecast["pv_estimate"],
+                            "pv_estimate10": forecast["pv_estimate10"],
+                            "pv_estimate90": forecast["pv_estimate90"],
                         }
                     )
 
-            _data = sorted(_data2, key=itemgetter("period_start"))
-            _fcasts_dict = {}
+            """
+            Add or update forecasts with the latest data.
+            """
 
+            # Load the forecast history.
+            forecasts = {}
             try:
-                for x in self._data['siteinfo'][site]['forecasts']:
-                    _fcasts_dict[x["period_start"]] = x
+                #for forecast in self._data['siteinfo'][site]['forecasts']:
+                #    forecasts[forecast["period_start"]] = forecast
+                forecasts = {forecast["period_start"]: forecast for forecast in self._data['siteinfo'][site]['forecasts']}
             except:
                 pass
-
-            _LOGGER.debug("Forecasts dictionary length %s", len(_fcasts_dict))
-
-            # Loop each site and its forecasts.
-            for x in _data:
-                itm = _fcasts_dict.get(x["period_start"])
-                if itm:
-                    itm["pv_estimate"] = x["pv_estimate"]
-                    itm["pv_estimate10"] = x["pv_estimate10"]
-                    itm["pv_estimate90"] = x["pv_estimate90"]
+            _LOGGER.debug("Forecasts dictionary length %s", len(forecasts))
+            forecasts_undampened = {}
+            try:
+                #for forecast in self._data_undampened['siteinfo'][site]['forecasts']:
+                #    forecasts_undampened[forecast["period_start"]] = forecast
+                forecasts_undampened = {forecast["period_start"]: forecast for forecast in self._data_undampened['siteinfo'][site]['forecasts']}
+            except:
+                pass
+            _LOGGER.debug("Undampened forecasts dictionary length %s", len(forecasts_undampened))
+
+            def update_forecast(forecasts: dict, period_start: dt, extant: dict, pv: float, pv10: float, pv90: float):
+                """Update an individual forecast entry."""
+                if extant: # Update existing.
+                    extant["pv_estimate"] = pv
+                    extant["pv_estimate10"] = pv10
+                    extant["pv_estimate90"] = pv90
+                else: # New forecast.
+                    forecasts[period_start] = {"period_start": period_start, "pv_estimate": pv, "pv_estimate10": pv10, "pv_estimate90": pv90}
+
+            def get_dampening_factor(site: str, z: dt):
+                """Retrieve a granular dampening factor."""
+                return self.granular_dampening[site][
+                    z.hour if len(self.granular_dampening[site]) == 24 else ((z.hour*2) + 1 if z.minute > 0 else 0)
```

Comment:
I think this should be:
else (z.hour*2 + (1 if z.minute > 0 else 0))

Reply:
Indeed